src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,182 @@
+        _ = jsonMessage ?? throw new ArgumentNullException(nameof(jsonMessage));
+
+        var assemblyPath = Path.Combine(assemblyFolderPath, assemblyFileName);
+
+        Document? document = null;
+        if (documentId is not null)
+        {
+            document = solution.GetDocument(documentId) ?? await solution.GetSourceGeneratedDocumentAsync(documentId, cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(document);
+        }
+
+        try
+        {
+            ICustomMessageHandlerWrapper? handler;
+            lock (_lockObject)
+            {
+                // Check if the assembly is already loaded.
+                if (!_extensions.TryGetValue(assemblyFolderPath, out var extension))
+                {
+                    var loadContext = new System.Runtime.Loader.AssemblyLoadContext(name: $"RemoteCustomMessageHandlerService assembly load context for {assemblyFolderPath}", isCollectible: true);
+                    loadContext.Resolving += ResolveExtensionAssembly;
+
+                    extension = new CustomMessageHandlerExtension(loadContext);
+                    _extensions[assemblyFolderPath] = extension;
+                }
+
+                if (!extension.Handlers.TryGetValue(typeFullName, out handler))
+                {
+                    var assembly = extension.AssemblyLoadContext.LoadFromAssemblyPath(assemblyPath);
+                    var type = assembly.GetType(typeFullName)
+                        ?? throw new InvalidOperationException($"Cannot find type {typeFullName} in {assemblyPath}.");
+
+                    handler = _customMessageHandlerFactory.Create(type);
+                }
+            }
+
+            var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+
+            var result = await handler.ExecuteAsync(message, document, solution, cancellationToken)
+                .ConfigureAwait(false);
+
+            var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+
+            return responseJson;
+        }
+        catch (Exception e)
+        {
+            Logger.Log(
+                FunctionId.CustomMessageHandlerService_HandleCustomMessageAsync,
+                $"Custom handler {assemblyPath} {typeFullName} error: {e}",
+                LogLevel.Error);
+            throw;
+        }
+
+        Assembly? ResolveExtensionAssembly(System.Runtime.Loader.AssemblyLoadContext context, AssemblyName assemblyName)
+        {
+            try
+            {
+                return _defaultLoadContext.LoadFromAssemblyName(assemblyName);
+            }
+            catch
+            {
+            }
+
+            var extensionAssemblyPath = Path.Combine(assemblyFolderPath, $"{assemblyName.Name}.dll");
+
+            // This will throw FileNotFoundException if the assembly is not found.
+            return context.LoadFromAssemblyPath(extensionAssemblyPath);
```

Comment:
Shouldn't we log an error if the assembly is not found?

Reply:
Currently we don't have access to the loghub LSP logger from OOP.  I think we probably need to make it available as I think that is generally the more appropriate place to log

Reply:
A failure in finding the assembly is logged the first time the handler is requested (see the `Error creating custom handler`... log). The exception is logged in full, so it should show that it's an issue with finding the assembly and which assembly was not found. The exception will also be bubbled up to the extension since not finding the assembly is likely an issue with how the extension was packaged.